### PR TITLE
Appdynamics

### PR DESCRIPTION
**Pull request by nkatre**
Support for AppDynamics service

* Downloads AppDynamics PHP agent

* Instruments php application and populates tier, node and app name for the application

* [ YES] I have viewed signed and have submitted the Contributor License Agreement

* [ YES] I have made this pull request to the `develop` branch

### CODE REVIEW

File: extensions/appdynamics/extension.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+             if (cups_svc == "appdynamics") or (cups_svc == "app-dynamics"):
+                _log.info("AppDynamics cups services detected.")
+                detected = True
+
+    if len(service_defs) > 0:
+        _log.debug("AppDynamics service detected.")
+        detected = True
+
+    if detected == True: 
+    	exit_code = os.system("echo preprocess_commands: AppDynamics agent configuration")
```
Comment (dmikusa-pivotal) [inline extensions/appdynamics/extension.py]: Why are you shelling out to `echo`?  You should almost certainly replace this with `_log.info("..")` or worst case `print ".."`.

File: extensions/appdynamics/extension.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        return [[ 'echo', '" in preprocess;"'],
+                ['env'],
+                [ 'chmod', ' -R 755 /home/vcap/app'],
```
Comment (dmikusa-pivotal) [inline extensions/appdynamics/extension.py]: You shouldn't hard code `/home/vcap/app`.  Use `$HOME` instead.

File: extensions/appdynamics/extension.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                detected = True
+
+    if len(service_defs) > 0:
+        _log.debug("AppDynamics service detected.")
+        detected = True
+
+    if detected == True: 
+    	exit_code = os.system("echo preprocess_commands: AppDynamics agent configuration")
+        return [[ 'echo', '" in preprocess;"'],
+                ['env'],
```
Comment (dmikusa-pivotal) [inline extensions/appdynamics/extension.py]: This is going to dump the full environment, which could expose sensitive info like passwords.  Remove this unless there's a very good reason.

File: extensions/appdynamics/extension.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                [ 'export', ' AD_CONTROLLER=`echo $VCAP_SERVICES | sed -e \'s/.*host-name.:.//g;s/\".*ssl-enabled.*//g\' `'],
+                [ 'export', ' AD_PORT=`echo $VCAP_SERVICES | sed -e \'s/.*port.:.//g;s/\".*account-access-key.*//g\' `'],
+                [ 'export', ' sslenabled=`echo $VCAP_SERVICES | sed -e \'s/.*ssl-enabled.:.//g;s/\".*.*//g\'`'],
+                [ 'if [ $sslenabled == \"true\" ] ; then export sslflag=-s ; fi; '],
+                [ 'echo sslflag set to $sslflag' ],
+                [ 'PATH=$PATH:./app/php/bin/ ./app/appdynamics/appdynamics-php-agent/install.sh $sslflag -i ./app/appdynamics/phpini -a=$AD_ACCOUNT_NAME@$AD_ACCOUNT_ACCESS_KEY $AD_CONTROLLER $AD_PORT $APP_NAME $APP_TIERNAME $APP_HOSTNAME' ],
+                [ 'cat', ' /home/vcap/app/appdynamics/phpini/appdynamics_agent.ini >> /home/vcap/app/php/etc/php.ini'],
+                [ 'cat', ' /home/vcap/app/appdynamics/phpini/appdynamics_agent.ini'],
+                [ 'echo', '"done preprocess"'],
+                ['env']]
```
Comment (dmikusa-pivotal) [inline extensions/appdynamics/extension.py]: Same.  This is going to dump the full environment which could contain sensitive info.  Remove.

File: extensions/appdynamics/extension.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+DEFAULTS = {
+'APPDYNAMICS_HOST': 'packages.appdynamics.com',
+'APPDYNAMICS_VERSION': '4.1.5.0',
+'APPDYNAMICS_PACKAGE': 'appdynamics-php-agent-x64-linux-{APPDYNAMICS_VERSION}.tar.bz2',
+'APPDYNAMICS_DOWNLOAD_URL': 'https://{APPDYNAMICS_HOST}/'
+                         'php/{APPDYNAMICS_VERSION}/{APPDYNAMICS_PACKAGE}',
+}
+
+class AppDynamicsInstaller(object):
```
Comment (dmikusa-pivotal) [inline extensions/appdynamics/extension.py]: New extensions should use [ExtensionHelper](https://github.com/cloudfoundry/php-buildpack/blob/master/lib/extension_helpers.py#L19) or [PHPExtensionHelper](https://github.com/cloudfoundry/php-buildpack/blob/master/lib/extension_helpers.py#L158).  This removes a lot of the boiler plate code in your extension.  See the [session extension](https://github.com/cloudfoundry/php-buildpack/blob/master/extensions/sessions/extension.py#L74) for an example.

File: extensions/appdynamics/extension.py (hunk 3 above)
Comment (flavorjones) [inline extensions/appdynamics/extension.py]: Agree. Do not log environment variables.

File: extensions/appdynamics/extension.py (hunk 1 above)
Comment (flavorjones) [inline extensions/appdynamics/extension.py]: Agree. Do not shell out unless absolutely necessary, as it introduces an opportunity for later developers to introduce remove execution vulnerabilities.

**Comment by flavorjones** [comment]
I'd prefer if this PR was focused on one thing, and not introducing unrelated changes such as the `CONTRIBUTING.md` file and the update to `PULL_REQUEST_TEMPLATE`. These are valuable but belong in a separate changeset.

**Comment by sesmith177** [comment]
@nkatre have you had a chance to review the comments from @flavorjones and @dmikusa-pivotal ?

Thanks.

**Comment by nkatre** [comment]
@sesmith177 I will make the required changes

**Comment by sesmith177** [comment]
@nkatre how has your progress on updating the PR gone?

**Comment by sesmith177** [comment]
@nkatre any updates here? Thanks.

**Comment by sesmith177** [comment]
@nkatre how is your progress on updating this PR going?

**Comment by nkatre** [comment]
@sesmith177 Apologize for the delay. We are reinvestigating on the changes required to support AppDynamics on php bp. I should have an update on this shortly.

**Pull request closed without merging**